File: create_listing.py
"""
Create a listing on OLX.ba (POST /listings) without running into the attribute 422s.

The gotcha: some categories require the `attributes` field and some reject it outright.
Sending an empty [] to a category that doesn't want attributes still fails. So the flow
is: look up the category's attributes first, then include the key only if it has any.

This example fills required attributes interactively for demo purposes. In a real
integration you'd map them from your own product data.

Run:
  python create_listing.py --token YOUR_TOKEN
"""
import argparse

import requests

API = "https://api.olx.ba"


def api(token):
    s = requests.Session()
    s.headers.update({"Accept": "application/json", "Authorization": "Bearer " + token})
    return s


def category_attributes(s, category_id):
    r = s.get(API + "/categories/%d/attributes" % category_id, timeout=20)
    return r.json().get("data") or []


def build_attributes(s, category_id):
    """
    Return (attributes_list, send_key).
    send_key is False when the category has no attributes, meaning you must NOT
    put the key in the payload at all.
    """
    attrs = category_attributes(s, category_id)
    if not attrs:
        return [], False

    filled = []
    for a in attrs:
        label = a.get("display_name") or a.get("name")
        options = a.get("options") or []
        if options:
            print("%s options: %s" % (label, options[:10]))
        value = input("value for '%s'%s: " % (label, " *" if a.get("required") else "")).strip()
        if value:
            filled.append({"id": int(a["id"]), "value": value})
    return filled, True


def create_listing(s, product, category_id):
    payload = {
        "type": "single",
        "title": product["title"],
        "short_description": product.get("short_description", product["title"]),
        "description": product["description"],  # basic HTML ok: <a href>, <br>
        "price": product["price"],
        "listing_type": "sell",
        "state": "new",
        "available": True,
        "category_id": category_id,
        "sku_number": product.get("sku", ""),
    }

    attributes, send_key = build_attributes(s, category_id)
    if send_key:
        payload["attributes"] = attributes  # only include when the category has attributes

    # brand / model only when present and known
    if product.get("brand_id"):
        payload["brand_id"] = product["brand_id"]
    if product.get("model_id"):
        payload["model_id"] = product["model_id"]

    r = s.post(API + "/listings", json=payload, timeout=30)
    if r.status_code >= 400:
        print("HTTP %d" % r.status_code)
        print(r.text)  # OLX puts the exact failing field in the errors object
        return None
    data = r.json().get("data") or r.json()
    listing_id = data.get("id")
    print("created listing id %s" % listing_id)
    return listing_id


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--token", required=True)
    args = ap.parse_args()

    s = api(args.token)

    product = {
        "title": "Example product",
        "description": "Line one.<br>Line two with a <a href=\"https://example.com\">link</a>.",
        "price": 25,
        "sku": "EXAMPLE-1",
    }
    category_id = int(input("category id to post into: ").strip())

    # after creating, you upload images separately via the image endpoints,
    # then publish. Left out here to keep the example focused.
    create_listing(s, product, category_id)


if __name__ == "__main__":
    main()

File: php/create_listing.php
<?php
/**
 * Create a listing (POST /listings) from PHP without hitting the attribute 422s.
 *
 * The catch: some categories require the `attributes` field and some reject it.
 * Sending [] to a category that doesn't want attributes fails too. So look up the
 * category attributes first and only include the key when there are any.
 *
 *   php create_listing.php
 */
require __DIR__ . '/olx_client.php';

// config
const TOKEN = 'YOUR_TOKEN';

$olx = new OlxClient(TOKEN);

/**
 * Build the attributes array for a category.
 * Returns [array $attributes, bool $sendKey]. When $sendKey is false the category
 * has no attributes and you must leave the key out of the payload entirely.
 */
function build_attributes(OlxClient $olx, int $categoryId, array $values): array
{
    $attrs = $olx->get("/categories/{$categoryId}/attributes") ?: [];
    if (!$attrs) {
        return [[], false];
    }
    $filled = [];
    foreach ($attrs as $a) {
        $id  = (int)$a['id'];
        $key = $a['name'] ?? (string)$id;               // map however your data is keyed
        $val = $values[$key] ?? $values[$id] ?? null;   // by attribute name or id
        if ($val !== null && $val !== '') {
            $filled[] = ['id' => $id, 'value' => (string)$val];
        }
    }
    return [$filled, true];
}

function create_listing(OlxClient $olx, array $product, int $categoryId, array $attributeValues = []): ?int
{
    $payload = [
        'type'              => 'single',
        'title'             => $product['title'],
        'short_description' => $product['short_description'] ?? $product['title'],
        'description'       => $product['description'],   // basic HTML ok: <a href>, <br>
        'price'             => $product['price'],
        'listing_type'      => 'sell',
        'state'             => 'new',
        'available'         => true,
        'category_id'       => $categoryId,
        'sku_number'        => $product['sku'] ?? '',
    ];

    [$attributes, $sendKey] = build_attributes($olx, $categoryId, $attributeValues);
    if ($sendKey) {
        $payload['attributes'] = $attributes;   // include only when the category has attributes
    }
    if (!empty($product['brand_id'])) {
        $payload['brand_id'] = (int)$product['brand_id'];
    }
    if (!empty($product['model_id'])) {
        $payload['model_id'] = (int)$product['model_id'];
    }

    $res = $olx->post('/listings', $payload);
    if ($res['status'] >= 400) {
        echo "HTTP {$res['status']}\n";
        // OLX returns the exact failing field under body.errors
        echo json_encode($res['body'], JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE) . "\n";
        return null;
    }
    $data = $res['body']['data'] ?? $res['body'];
    $id   = $data['id'] ?? null;
    echo "created listing id {$id}\n";
    return $id ? (int)$id : null;
}

// demo
$product = [
    'title'       => 'Example product',
    'description' => "Line one.<br>Line two with a <a href=\"https://example.com\">link</a>.",
    'price'       => 25,
    'sku'         => 'EXAMPLE-1',
];

// after this you'd upload images via the image endpoints and publish; left out to stay focused
create_listing($olx, $product, 2369, ['Stanje' => 'Novo']);
